### PR TITLE
Support injecting partial failures in mocks

### PR DESCRIPTION
**Pull request by leesio**
# What?
When writing code which combines operations with `op.Add`, it's important to consider data consistency in the case of partial failures.

This PR offers a way to test those cases, providing a way to generate a special `context.Context` which, when provided to `mockMultiOp.RunWithContext` will inject errors before different operations.

Two error injection strategies are provided, but there's scope for alternative strategies.

https://github.com/monzo/gocassa/blob/da3fa14f92b63e973bf706166b395649cf559a74/mock.go#L1024-L1062

### CODE REVIEW

File: mock.go [hunk 1]
```diff
@@ -909,3 +913,31 @@ func assignRecords(m map[string]interface{}, record map[string]interface{}) erro
 
 	return nil
 }
+
+type mockContextKey string
+
+var partialFailureContextKey mockContextKey = "partial_failure"
+
+type ShouldFailFunc func(opIdx, opCount int) error
```
Comment (krishishah) [inline mock.go:921]: I think some docs would be helpful on the following:
 * what `opIdx` and `opCount` mean
 * gotchas e.g. `idx` is 0 indexed but `opCount` isn't so `opCount - 1` will give the user the final op not the second to last one.
 * What purposes have we designed `PartialFailureContext` to satisfy?

Comment (krishishah) [reply mock.go:921]: Should we add in example test cases? (or at least test that this behaves as we expect)

**Review by krishishah — COMMENTED** [review]
Thanks @leesio! Really happy with this in its current state. 

Left a couple of minor comments

File: mock.go [hunk 2]
```diff
@@ -909,3 +913,150 @@ func assignRecords(m map[string]interface{}, record map[string]interface{}) erro
 
 	return nil
 }
+
+type mockContextKey string
+
+var errorInjectorContextKey mockContextKey = "failure_strategy"
```
Comment (kjgorman) [inline mock.go:919]: should we update this key text to match the change from 'failure strategy' to 'error injector'?

File: mock.go [hunk 3]
```diff
@@ -909,3 +913,150 @@ func assignRecords(m map[string]interface{}, record map[string]interface{}) erro
...
+		}
+	}
+	return nil
+}
+
+// PartialFailureContext returns a context which when passed to
+// mockMultiOp.RunWithContext(...), will inject an error before one of the
+// operations to simulate a partial failure in the query execution.
+// The ErrorInjector determines when in the sequence the error is injected
+func PartialFailureContext(parent context.Context, strategy ErrorInjector) context.Context {
```
Comment (kjgorman) [inline mock.go:947]: as above perhaps s/strategy/injector

**Review by kjgorman — COMMENTED** [review]
could we add a couple of unit tests (e.g. one that covers 'nth failure') to show/prove usage along with the example?

**Review by kjgorman — APPROVED** [review]

**Pull request merged**